### DISCUSSION TITLE
class Conversations:  How to receive all the photos that a user has sent in one message?

### DISCUSSION THREAD

**Discussion by Kromvel-X** [nutgram/nutgram#676] [Q&A] [answered]
Hello!
Thank you for developing such a wonderful library! :)

Please tell me, is it possible in the Conversation class to receive multiple photos sent by the user in one message?

I know that in the base Nutgram class, you can get all the photos sent in one message using handlers:

1) onMessageType:
$bot->onMessageType(MessageType::PHOTO, function (Nutgram $bot) {
$photo = end($bot->message()->photo);
$bot->getFile($photo->file_id)?->save('./');
});

2) onPhoto:
$bot->onMessageType(MessageType::PHOTO, function (Nutgram $bot) {
    $photo = end($bot->message()->photo);
    $bot->getFile($photo->file_id)?->save('./');
    $bot->sendMessage('Photo saved!');
});

But in the Conversation class, if I send several photos in one message, I can only get one - the first one.
As far as I understand, even if several photos were sent in one message, they are actually still uploaded in separate messages and it turns out that 1 photo = 1 message.

I extended the Conversation class and created a getPhoto method, which receives the photo and saves its URL in a property:

    /**
     * Get photo
     *
     * @param Nutgram $bot
     * @return boolean
     */
    public function getPhoto(Nutgram $bot):bool
    {
        // Getting the message type
        $MessageType = $bot->message()->getType()->value;
        // Check if the message is a photo
        if( $MessageType !== MessageType::PHOTO->value){
            return false;
        }
        $photo = end($bot->message()->photo);
        // Getting a photo file
        $file = $bot->getFile($photo->file_id);
        // Get the url of the photo
        $photoUrl = $file->url();
        // Write the url of the photo to the property for storage
        $this->userData['photos'][] = $photoUrl;
        return true;
    }

I call this method on the step that follows after the user uploads photos:

    /**
     * Step photos
     *
     * @param Nutgram $bot
     * @return void
     */
    public function stepPhotos(Nutgram $bot):void
    {
        $this->userData['location'] = $this->getLocation($bot);
        $bot->sendMessage('Upload a photo of the salon');
        $this->next('stepEmail');
    }

    /**
     * Step email
     *
     * @param Nutgram $bot
     * @return void
     */
    public function stepEmail(Nutgram $bot):void
    {
        $photo = $this->getPhoto($bot);
        if(!$photo){
            $this->stepPhotos($bot);
            return;
        }
        $bot->sendMessage('Enter the email address of the salon');
        $this->next('stepPhone');
    }

As a result, I only get the first photo, out of all to have been sent in one message.

Please tell me what I'm missing and where I made a mistake, and is it possible at all to receive all the photos that were sent in one message in the Conversation class?

**Comment by artcd** [comment]
I'm also interested in how to get all the files during a conversation🙄

**Comment by artcd** [comment]
Use this code to upload photos in conversation mode ⬇️
```php
foreach($bot->getUpdates() as $update)
{
	if(!empty($update->getMessage()->photo))
	{
		$photo      = end($update->getMessage()->photo);
		$link       = $bot->getFile($photo->file_id)->url();
		$filename   = pathinfo($link, PATHINFO_BASENAME);
		
		Storage::disk('public')->put($filename, file_get_contents($link));
	}
}
```

**Comment by artcd** [comment]
Unfortunately my method only works if you run the bot locally via console command:
```
php artisan nutgram:run
```
If you use a webhook mode it doesn't work😢

Does anyone have any ideas on how to get files from a user in conversation mode?

**Reply by Kromvel-X** [reply, marked as answer]
I have come up with another solution. It will be interesting to hear your opinion.
As far as I understand, the problem is that when you send multiple photos in one message, they are still sent as separate messages. So in the next method, which is called after sending a message with photos, you need to check the type of the received message and if it is of type ‘photo’ you need to run the get photos method again. This way you can process all the photos sent in one message.

Here is a complete example:

```php
/**
 * Request a photo of the salon
 * Save the answer to the previous request - the geolocation of the salon - into a variable.
 * Move to the next step of the dialogue: stepRequestEmail * 
 * @param NslabBot $bot - bot instance
 * @return void
 */
public function stepRequestPhotos(NslabBot $bot):void
{
    $location = $this->getLocation($bot);
    if(!$location){
        $bot->sendMessage('Error: problem with location detection, try again');
        $this->stepRequestLocation($bot);
        return;
    }

    $this->salon->setLocation($location);

    $this->deleteKeyboard($bot);
    $bot->sendMessage('Upload a photo of the interior');
    $this->next('stepRequestEmail');
}

/**
 * Request the email address of the salon
 * Save the answer to the previous request - the photo of the salon - into a variable.
 * Move to the next step of the dialogue: stepRequestPhone
 * 
 * @param NslabBot $bot - bot instance
 * @return void
 */
public function stepRequestEmail(NslabBot $bot):void
{
    $photo = $this->getPhoto($bot);
    if(!$photo){
        $this->stepRequestPhotos($bot);
        return;
    }
    $bot->sendMessage('Email the salon');
    $this->next('stepRequestPhone');
}
```
This method checks the type of the received message ```if( $MessageType === MessageType::PHOTO->value)``` and if we have received a photo, we run the method of processing this photo ```$this->getPhoto($bot)``` and then call the next dialogue method.

```php
/**
 * Request the salon's phone number
 * Save the answer to the previous request - email address of the salon or a photo of the salon - into a variable.
 * Go to the next step of the dialogue: stepRequestContactPersonName or run the repeated getPhoto photo retrieval.
 *
 * @param NslabBot $bot - bot instance
 * @return void
 */
public function stepRequestPhone(NslabBot $bot):void
{
    $message = $bot->message();
    $MessageType = $message->getType()->value;
    if( $MessageType === MessageType::PHOTO->value){
        $this->getPhoto($bot);
        $this->next('stepRequestPhone');
        return;
    }

    $messageText = $message->text;
    $email = $this->validateEmail($messageText);
    if(!$email){
        $bot->sendMessage('Error: invalid email address, try again');
        $this->next('stepRequestPhone');
        return;
    }
    $this->salon->setEmail($messageText);
    $bot->sendMessage('Give me the salon's phone number');
    $this->next('stepRequestContactPersonName');
}
```

**Reply by artcd** [reply]
Please to help fix this code:
```php
<?php

namespace App\Telegram\Conversations;

use SergiX44\Nutgram\Nutgram;
use Illuminate\Support\Facades\Storage;
use Psr\SimpleCache\InvalidArgumentException;
use SergiX44\Nutgram\Conversations\Conversation;
use SergiX44\Nutgram\Telegram\Types\Keyboard\InlineKeyboardButton;
use SergiX44\Nutgram\Telegram\Types\Keyboard\InlineKeyboardMarkup;

class Uploader extends Conversation
{
    protected int $message_id;
    protected string $text = 'Загрузите ваши файлы ниже 🔽';
    protected InlineKeyboardMarkup|null $keyboard;
    protected bool $finish = false;
    protected array $files;

    /**
     * @throws InvalidArgumentException
     */
    public function start(Nutgram $bot): void
    {
        // Клавиатура
        $this->keyboard = InlineKeyboardMarkup::make()
            ->addRow(
                InlineKeyboardButton::make(
                    text: '❌ Отмена',
                    callback_data: 'cancel'
                )
            );

        if(empty($this->message_id))
        {
            // Печатаем сообщение
            $this->message_id = $bot->sendMessage(
                text: $this->text,
                reply_markup: $this->keyboard
            )->message_id;
        }
        else
        {
            // Редактируем сообщение
            $bot->editMessageText(
                text: $this->text,
                //reply_markup: $this->keyboard
            );
        }

        // Шаг загрузки файлов
        $this->beforeStep($bot);
        $this->finish ? $this->end() : $this->next('upload');
    }

    /**
     * @throws InvalidArgumentException
     */
    public function upload(Nutgram $bot): void
    {
        // Если нажали по кнопке
        if($bot->isCallbackQuery())
        {
            // Получаем экшен кнопки
            $data = $bot->callbackQuery()->data;

            // Если нажали отмена, то закрываем диалог
            if($data == 'cancel')
            {
                // Текст
                $this->text = 'Вы отменили загрузку файлов';

                // Клавиатура
                //$this->keyboard = null;

                // Финиш
                $this->finish = true;

                // Вернёмся в начало
                $this->start($bot);
            }
        }
        else
        {
            // Тип вложений
            if(!empty($bot->message()->photo))
            {
                $attach = end($bot->message()->photo);
                $type   = 'photo';
            }
            elseif(!empty($bot->message()->document))
            {
                $attach = $bot->message()->document;
                $type   = 'document';
            }
            elseif(!empty($bot->message()->video))
            {
                $attach = $bot->message()->video;
                $type   = 'video';
            }

            // Загрузка файлов
            if(!empty($attach) && !empty($type))
            {
                // Информация о файле
                $file = $bot->getFile($attach->file_id);

                // Расширение файла
                $ext = pathinfo($file->file_path, PATHINFO_EXTENSION);

                // Путь для сохранения
                $path = Storage::disk('attachments')->path("{$file->file_unique_id}.{$ext}");

                // Загружаем файлы на сервер
                $attach->download($path);

                $this->files[] = $file;
            }
        }
    }

    /**
     * @throws InvalidArgumentException
     */
    public function preview(Nutgram $bot): void
    {
        $bot->sendMessage(
            text: json_encode($this->files, JSON_PRETTY_PRINT)
        );

        // Закрываем диалог
        $this->end();
    }
}
```

**Reply by Kromvel-X** [reply]
I can take a closer look at your code later today, I don't have time for it now, but here are my thoughts on uploading files:
You need to implement 3 methods in the Uploader class:
1) method_1 : file upload request step to users:
    -> for example you send a message to a user asking them to upload files.
    -> call method_2 ($this->next('method_2');)

2) method_2 (step that is executed after method_1 ):
    -> check for the type of the received message: 
        1) if file :
            - then call method_3 ($this->method_3();)
            - call the method_2 ($this->next('method_2');)
             - return from method_2 (return;)
        2) if other type of message - continue executing method_2 

3) method_3:  processing sent files. It will be executed inside method_2. 
    -> save file.

**Reply by Kromvel-X** [reply]
@artcd I have some free time now, so if you need more clarification then write me I will try to help you.

**Reply by artcd** [reply]
@Kromvel-X, here is my new code:
```php
<?php

namespace App\Telegram\Conversations;

use SergiX44\Nutgram\Nutgram;
use Psr\SimpleCache\InvalidArgumentException;
use SergiX44\Nutgram\Conversations\Conversation;
use SergiX44\Nutgram\Telegram\Types\Keyboard\InlineKeyboardButton;
use SergiX44\Nutgram\Telegram\Types\Keyboard\InlineKeyboardMarkup;

class Uploader extends Conversation
{
    protected array $userData;

    /**
     * @throws InvalidArgumentException
     */
    public function start(Nutgram $bot):void
    {
        // Sending a message
        $bot->sendMessage(
            text: 'Upload your files 🔽'
        );

        // File upload step
        $this->next('upload');
    }

    /**
     * @throws InvalidArgumentException
     */
    public function upload(Nutgram $bot):void
    {
        if(!$this->getPhoto($bot))
        {
            $this->start($bot);
            return;
        }

        $bot->sendMessage(
            text: 'Files uploaded',
            reply_markup: InlineKeyboardMarkup::make()
                ->addRow(
                    InlineKeyboardButton::make(
                        text: 'Approve',
                        callback_data: 'approve'
                    )
                )
        );

        // Preview step
        $this->next('preview');
    }

    /**
     * @throws InvalidArgumentException
     */
    public function preview(Nutgram $bot):void
    {
        // Checking a message for a photo
        if($bot->message()->photo){
            $this->getPhoto($bot);
            $this->next('preview');
            return;
        }

        $bot->sendMessage(
            text: json_encode($this->userData, JSON_PRETTY_PRINT)
        );
    }

    public function getPhoto(Nutgram $bot):bool
    {
        // Check if the message is a photo
        if(!$bot->message()->photo){
            return false;
        }

        // Getting the biggest photo
        $photo = end($bot->message()->photo);

        // Getting a photo file
        $file = $bot->getFile($photo->file_id);

        // Get the url of the photo
        $photoUrl = $file->url();

        // Write the url of the photo to the property for storage
        $this->userData['photos'][] = $photoUrl;

        return true;
    }
}
```
It works fine but you need to perform some action to read the array of photos.
But I would like to go to the preview step immediately after loading all the files.

**Reply by Kromvel-X** [reply]
@artcd 
Honestly, I’m not sure how to implement it in Conversation mode. If I understand the structure of Conversation correctly, here’s the situation:

1. Conversation is a set of sequential steps where the user interacts with the chatbot.
2. Moving from one step to the next happens when the user sends a message to the chatbot.

_When the user sends a media group with photos, all the photos are sent as separate messages._

**Conclusion:**  
I could be wrong, but it seems to me that after the user sends a message containing a media group with images, it’s impossible to proceed without an additional **upload** method, which would require the user to send an extra message (in your case, pressing the 'Approve' button - (InlineKeyboardButton)). This way, we create a starting point that helps understand that all images have been uploaded and processed, and we can move on to the next step, **preview**.

If you have thoughts on this I'd love to hear them. 

It would be interesting to hear @Lukasss93's opinion on this.

**Reply by Lukasss93** [reply]
This is the way.
